File: src/magma/app/compile/rule/SuffixRule.java
package magma.app.compile.rule;

import magma.api.Err;
import magma.app.compile.GenerateError;
import magma.app.compile.Node;
import magma.app.compile.ParseError;

public final class SuffixRule implements Rule {
    private final Rule child;
    private final String suffix;

    public SuffixRule(Rule child, String suffix) {
        this.child = child;
        this.suffix = suffix;
    }

    @Override
    public RuleResult<Node, ParseError> parse(String input) {
        if (!input.endsWith(suffix))
            return RuleResult.RuleResult(Err.Err(new ParseError("Suffix not present '" + suffix + "'", input)));
        var name = input.substring(0, input.length() - suffix.length());
        return child.parse(name);
    }

    @Override
    public RuleResult<String, GenerateError> generate(Node node) {
        return child.generate(node).wrapValue(inner -> inner + suffix);
    }
}

File: src/ts/magma/app/compile/rule/SuffixRule.ts
import { Rule } from "../../../../magma/app/compile/rule/Rule";
import { Option } from "../../../../magma/api/option/Option";
import { None } from "../../../../magma/api/option/None";
import { Strings } from "../../../../magma/api/text/Strings";
export class SuffixRule<T> implements Rule<T> {
	suffix: string;
	childRule: Rule<T>;
	constructor (suffix: string, childRule: Rule<T>) {
		this.suffix = suffix;
		this.childRule = childRule;
	}
	lex(input: string): Option<T> {
		if (!input/*string*/.endsWith(this.suffix)/*unknown*/){
			return new None<T>()/*unknown*/;
		}
		let length = Strings.length(input)/*unknown*/;
		let length1 = Strings.length(this.suffix)/*unknown*/;
		let content = Strings.sliceBetween(input, 0, length - length1)/*unknown*/;
		return this.childRule.lex(content)/*unknown*/;
	}
}
